### ISSUE TITLE
`itemToString` - strange arguments with multiselect

### ISSUE THREAD

**Issue by jamesplease** [paypal/downshift#395]
* `downshift` version: 1.30.1
* `node` version: 9.x
* `npm` (or `yarn`) version: 5.x

Relevant code or config

See the multiselect example [from the README](https://codesandbox.io/s/W6gyJ30kn).

What you did:

I updated `itemToString` to be:

```
itemToString={i => {
  console.log('hello', i);
  return i.name;
}}
```

What happened:

Sometimes, `itemToString` is called with an object, and other times, an Array.

![image](https://user-images.githubusercontent.com/2322305/37842743-2fa3bbf6-2e80-11e8-92ea-97abdcc1bcc1.png)

Reproduction repository:

[The multiselect example from the README](https://codesandbox.io/s/W6gyJ30kn)

Problem description:

It seems unexpected that it is sometimes called with the individual items, and other times an array of the selected items. The documentation does not cover this, and sometimes it can cause unexpected results.

For instance, if an input was used in that `multiselect` example, instead of a button, the `itemToString` function would cause the string `undefined` to render. I can make an example demonstrating this if it would be valuable.

Suggested solution:

I'm not sure! I haven't looked under-the-hood to see why it behaves this way; it just struck me as odd.

---

I am willing to make a PR if there are code changes that you think should be made. Otherwise, I am open to updating the documentation with a guide that describes this behavior in greater detail, or solutions on how to deal with it.

Thanks for reading!

**Comment by kentcdodds** [comment]
Hi @jamesplease! Thanks for bringing this up!

Yeah, this is tricky. First of all we're calling it ourselves with a single item in a few places in that example. I've updated it to not do that anymore. Secondly, [in some places](https://github.com/paypal/downshift/blob/ee64fd8a58d5e9e3b80a55a172a88072e9f72a2a/src/downshift.js#L819) `itemToString` will be called on an `item` and in [other places](https://github.com/paypal/downshift/blob/ee64fd8a58d5e9e3b80a55a172a88072e9f72a2a/src/downshift.js#L129) it'll be called on the `selectedItem`. I'm not sure how to handle this. I'm thinking the best thing we can do is update the example (which I've done). I don't know if it's necessary to document this (or even where those docs would go). I've added a comment about it in the example. That should hopefully be enough.

Thanks again!

**Comment by jamesplease** [comment]
Thanks @kentcdodds ! The updated example is 👌. I think that is a good solution.

---

As an aside, I wonder if a Downshift that was built to support multi-select by default would have a slightly cleaner API. The current multi select works, but it feels like it is tacked onto the primary use case of a single selected item. I haven't put too much thought into the idea – it may not be possible, or that API may be weird in other ways.

Anyway, thanks again!

**Comment by jamesplease** [comment]
> Heads up: this reply is long. I respond to the above comment at the end, but first...

I have been playing with the code in the updated [multi-select example](https://codesandbox.io/s/W6gyJ30kn), which is:

```js
itemToString = item => {
   if (!item) {
    return ''
  } else if (Array.isArray(item)) {
    return item.map(i => i.name).join(', ')
  } else {
    return item.name
  }
}
```

This solves some problem with `itemToString`, but the API will return inconsistent results in other areas with this code. For instance, `onInputValueChange` will pass you strings sometimes, and arrays other times.

For instance, you may get these values passed:

```
"L"
"Lo"
"Lor"
"Lord of the Rings"
["Lord of the Rings"]
```

In my opinion, if it is passing me an array, then it cannot be passing me the input value, because an input's value should always be a string.

There are two solutions I can think of to this. The first is to add an additional conditional within this callback that checks for an array or a string. I think there are two problems with this solution:

1. it sets an unfortunate precendent: always checking for a string or an array. It seems reasonable to expect that the library should be able to handle these checks for the consumers

2. the code that developers must write is no longer very expressive. `itemToString` and `onInputValueChanged` are amazing methods: the names tell you exactly what you are being passed and, in the case of `itemToString`, it tells you what it expects you to return. You probably do not even need to read the documentation for these methods to use them correctly for the single select use case!
  However, that expressiveness is lost for the multi select case. What is an amazing API for single selections instead feels like a _super hacky_ API.
  The problem with that is that it is difficult for developers to feel confident about hacky-feeling things. I love using Downshift for single selection, but I am far less confident that any of the code for multi selection will work as expected in every situation.

Should a developer decide to go with the array-checking approach, a descriptive argument name for the callback is pretty unusual, I think: `inputValueOrStringifiedSelectedItems`. Or, they could go with `inputValue` and leave a comment that it is sometimes _not_ the input value, but that does not seem great.

To avoid the hackiness within `onInputValueChanged`, another solution is to specify your `itemToString` as the following:

```js
itemToString = item => {
   if (!item || Array.isArray(item)) {
    return ''
  } else {
    return item.name
  }
}
```

I haven't tested the behavior of this kind of `itemToString` for every Downshift API, but in regards to `onInputValueChange`, this implementation ensures two things:

1. `onInputValueChanged` will always be passed a string, which is good, I think
2. `onInputValueChanged` will only pass to you what the user has, well, typed into the input. If they make some selections, but the input is empty, then an empty string will be passed to you (because there is nothing in the input)

This is an improvement, I think, because `onInputValueChanged` retains that same degree of expressiveness that you get for the single selection use case – which is great!

Of course, the problem with this `itemToString` is that it still doesn't make any sense to read. The name `itemToString` suggests to me that it will always be passed a single item, and it expects you to convert that item to a string. It is strange that it will pass you an array sometimes, and the implications of what happens when you handle that array are simply not clear in the API.

For instance, it doesn't seem clear to me that mapping the array to a string of values would change the behavior of `onInputValueChanged`.

---

My current belief is that a Downshift that supported multiple selection _first_ would likely have a more expressive API for both the multiple selection and single selection use cases.

Just to throw some code out there, I think users would need to do two patterns to support the single-selection use case in a multiple selection Downshift world.

The first is related to getting the selected item. I think that that code would look like this:

```js
// we only allow the user to select one item, so the selected item
// is always the first item in the selected items list
const selectedItem = selectedItems[0];
```

When it comes to updating the value, I expect that the code would look something like:

```js
// because the library was built to support multiple selection
// first, we must pass an array. But we just override the existing array
// with an array with _just_ the newly-selected item
changeValue([selectedItem]);
```

Above, a few examples were given of what it looks like to support multiple selection in Downshift, but the tl;dr is that it looks like this:

```js
youWillBePassedOneThing(thingOrArrayOfThings) {
  if (!thingOrArrayOfThings) {
    // Situation 1
  } else if (Array.isArray(thingOrArrayOfThings)) {
    // Situation 2
  } else {
    // Situation 3
  }
}
```

That alone is weird, but it is worse that the way that you handle the array situation can cause unexpected changes to the behavior of other props, too.

> I'll bet you could build the kind of component you're hoping about with just a few dozen lines of code. I hope you do :)

This may be true, but I am not sure that it would be possible to make a really great multiple selection downshift API from the current Downshift. It is usually easier to go from a bulk API to a singular API, and I think that would apply in this situation, too.

---

tl;dr thoughts:

- I do not think that Downshift's current API is a great tool to use for multiple selection dropdowns (but it is **amazing** for the singular use case! I cannot stress that enough, Downshift is **amazing** @kentcdodds !!)
- I think it is reasonable to expect that a multiple selection dropdown would work reasonably well for singular use cases as well
- If the above is not true, then it would likely be easier to make a nice singular API from the multiple selection API than it would be to make a good multiple selection API from Downshift's current singular API

**Comment by kentcdodds** [comment]
Alrighty, you twisted my arm. Give the example a look now: https://codesandbox.io/s/W6gyJ30kn

You'll see I re-implemented this to create a `MultiDownshift` component in a `./multi-downshift.js` file that is 90 lines long (a couple dozen lines :wink:). It could use some work on supporting control props and composing some things together (like the `stateReducer`), but for the most part this is what I envisioned people do for common cases when using downshift. I'm honestly surprised nobody's done much of this kind of composition yet. Downshift is called a primitive for a reason. I have a dream for someone to reimplement the entire react-select API in a few hundred lines using downshift under the covers.

Let me know what you think of this. I'd love to have something like this published to npm :)

**Comment by jamesplease** [comment]
Thank you for taking the time to put together some code as a reply @kentcdodds ! I don't have time to look at it _right this second_, but I plan to in the next few days or weeks. I will write back here once I do get around to it!

> I'd love to have something like this published to npm :)

:v: Agreed!
